src/tsserver/server.ts

```diff
@@ -648,375 +22,70 @@ namespace ts.server {
-            if (!watchedFile) {
-                return;
-            }
-
-            fs.stat(watchedFile.fileName, (err, stats) => {
-                if (err) {
-                    if (err.code === "ENOENT") {
-                        if (watchedFile.mtime.getTime() !== 0) {
-                            watchedFile.mtime = missingFileModifiedTime;
-                            watchedFile.callback(watchedFile.fileName, FileWatcherEventKind.Deleted);
-                        }
-                    }
-                    else {
-                        watchedFile.callback(watchedFile.fileName, FileWatcherEventKind.Changed);
-                    }
-                }
-                else {
-                    onWatchedFileStat(watchedFile, stats.mtime);
-                }
-            });
-        }
-
-        // this implementation uses polling and
-        // stat due to inconsistencies of fs.watch
-        // and efficiency of stat on modern filesystems
-        function startWatchTimer() {
-            // eslint-disable-next-line no-restricted-globals
-            setInterval(() => {
-                let count = 0;
-                let nextToCheck = nextFileToCheck;
-                let firstCheck = -1;
-                while ((count < chunkSize) && (nextToCheck !== firstCheck)) {
-                    poll(nextToCheck);
-                    if (firstCheck < 0) {
-                        firstCheck = nextToCheck;
-                    }
-                    nextToCheck++;
-                    if (nextToCheck === watchedFiles.length) {
-                        nextToCheck = 0;
-                    }
-                    count++;
-                }
-                nextFileToCheck = nextToCheck;
-            }, interval);
-        }
-
-        function addFile(fileName: string, callback: FileWatcherCallback): WatchedFile {
-            const file: WatchedFile = {
-                fileName,
-                callback,
-                mtime: sys.fileExists(fileName)
-                    ? getModifiedTime(fileName)
-                    : missingFileModifiedTime // Any subsequent modification will occur after this time
-            };
-
-            watchedFiles.push(file);
-            if (watchedFiles.length === 1) {
-                startWatchTimer();
-            }
-            return file;
-        }
-
-        function removeFile(file: WatchedFile) {
-            unorderedRemoveItem(watchedFiles, file);
-        }
+    export interface StartInput {
+        args: readonly string[];
+        logger: Logger;
+        cancellationToken: ServerCancellationToken;
+        serverMode: LanguageServiceMode | undefined;
+        unknownServerMode?: string;
+        startSession: (option: StartSessionOptions, logger: Logger, cancellationToken: ServerCancellationToken) => void;
     }
+    function start({ args, logger, cancellationToken, serverMode, unknownServerMode, startSession: startServer }: StartInput, platform: string) {

```

Comment:
This seems like a convoluted way to share this initialization code.  For example, `serverMode` is extracted from the `StartInput` and then passed back into `startServer`.  It might be tidier to extract some initialization helpers (e.g. `logHeader`) and then have each code path (i.e. web and node) call them separately.  Such an exercise might also reveal bits that shouldn't be shared (e.g. why does the web server need to print out the node version in its header?).